### PR TITLE
BAEL-4779 Setting TTL for Redis in Spring Framework

### PR DESCRIPTION
**Pull request by sharanm**

### CODE REVIEW

File: persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        // key has been deleted
+        assert bookRepository.get(book.getTitle()) == null;
+    }
+    
+    @Test
+    public void testLibrarianInfo() throws InterruptedException {
+        Librarian librarian = new Librarian(5L, "Joe");
+        librarianRepository.save(librarian);
+        
+        assert librarianRepository.findById(librarian.getId()).get().getName()
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java:79]: Can we use JUnit assertions here or assertJ - these aren't really assertions ideal for testing.

File: persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+          .equals(librarian.getName());
+        
+        Thread.sleep(6000);
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java:82]: This adds a fairly long wait to the automation tests. Could we maybe set these tests to `@Ignore` so as not to slow down the whole Integration test build.

File: persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        // key has been deleted
+        assert librarianRepository.findById(librarian.getId()).isPresent() == false;
+    }
+
+    
+    private Book getDummyBook(String title, String lang) {
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java:88]: If this just calls a constructor, might it be cleaner not to have this method at all?

File: persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        Book book = new Book(title, lang);
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java:89]: Why do we need a temporary variable here if we're returning it immediately? The `return` statement would do the whole job on one line!

File: persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import com.baeldung.spring.data.redis_ttl.repository.LibrarianRepository;
+
+import redis.embedded.RedisServerBuilder;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT, classes = LibraryApplication.class)
+@DirtiesContext(classMode = ClassMode.BEFORE_CLASS)
+public class LibraryIntegrationTest {
+    
+    private static redis.embedded.RedisServer redisServer;
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java:27]: Do we need to fully qualify this type?

File: persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    
+    @BeforeClass
+    public static void startRedisServer() throws IOException {
+        redisServer = new RedisServerBuilder().port(6379).setting("maxmemory 256M").build();
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java:31]: This runs the risk of port clashing with anything else running on Jenkins. Perhaps this test shouldn't be run at all in the Baeldung CI Build?

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/repository/GatekeeperRepository.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package com.baeldung.spring.data.redis_ttl.repository;
+
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.stereotype.Repository;
+
+import com.baeldung.spring.data.redis_ttl.entity.Gatekeeper;
+
+@Repository
+public interface GatekeeperRepository extends CrudRepository<Gatekeeper, Long>{
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/repository/GatekeeperRepository.java:9]: missing space before `{`

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/repository/BookRepository.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    public Book get(String title) {
+        return redisTemplate.opsForValue().get(title);
+    }
+    
+    public void save(Book book) {
+        redisTemplate.opsForValue().set(book.getTitle(), book);
+        
+        int ttlValue = 5; //secs 
+        if(book.getLanguage().toLowerCase().equals("english"))
+            ttlValue = 10;
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/repository/BookRepository.java:31]: missing `{}` around if statement's expression

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/repository/BookRepository.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.stereotype.Repository;
+
+import com.baeldung.spring.data.redis_ttl.entity.Book;
+
+@Repository
+public class BookRepository{
+
+    private RedisTemplate<String, Book> redisTemplate;
+    
+    @Autowired
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/repository/BookRepository.java:16]: `@Autowired` isn't needed on constructors in spring

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/repository/BookRepository.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.concurrent.TimeUnit;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.redis.core.RedisTemplate;
+import org.springframework.stereotype.Repository;
+
+import com.baeldung.spring.data.redis_ttl.entity.Book;
+
+@Repository
+public class BookRepository{
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/repository/BookRepository.java:12]: Missing space before `{`

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/entity/Subscriber.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.io.Serializable;
+
+import org.springframework.data.annotation.Id;
+import javax.persistence.Entity;
+
+@Entity
+public class Subscriber implements Serializable{
+    private static final long serialVersionUID = 1L;
+    @Id private Long id;
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/entity/Subscriber.java:11]: Let's put a line break after `@Id` and let's separate fields from each other with blank lines.

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/entity/Librarian.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.io.Serializable;
+
+import org.springframework.data.annotation.Id;
+import org.springframework.data.redis.core.RedisHash;
+
+@RedisHash(value="librarian", timeToLive = 5)
+public class Librarian implements Serializable{
+    private static final long serialVersionUID = 1L;
+    @Id private Long id;
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/entity/Librarian.java:11]: Let's put a new line after `@Id` and separate fields with new lines

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/entity/Librarian.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
+package com.baeldung.spring.data.redis_ttl.entity;
+
+import java.io.Serializable;
+
+import org.springframework.data.annotation.Id;
+import org.springframework.data.redis.core.RedisHash;
+
+@RedisHash(value="librarian", timeToLive = 5)
+public class Librarian implements Serializable{
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/entity/Librarian.java:9]: Missing space before `{`

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/entity/Gatekeeper.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
+package com.baeldung.spring.data.redis_ttl.entity;
+
+import java.io.Serializable;
+import javax.persistence.Entity;
+import org.springframework.data.annotation.Id;
+
+@Entity
+public class Gatekeeper implements Serializable{
+    private static final long serialVersionUID = 1L;
+    @Id private Long id;
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/entity/Gatekeeper.java:10]: Let's put a new line after @Id and separate fields with new lines

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/entity/Book.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
+package com.baeldung.spring.data.redis_ttl.entity;
+
+import java.io.Serializable;
+
+public class Book implements Serializable{
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/entity/Book.java:5]: Missing space before `{`

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/configuration/RedisConfig.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+        // Setting the expiration time of the cache
+        Map<String, RedisCacheConfiguration> configurations = new HashMap<>();
+        configurations.put("subscriber", shortDurationCache);
+        configurations.put("gatekeeper", longDurationCache);
+
+        // RedisCacheManager redisCacheManager =
+        // RedisCacheManager.builder(connectionFactory)
+        // .transactionAware()
+        // .cacheDefaults(config).build();
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/configuration/RedisConfig.java:47]: Please explain in a comment why this code is commented out - I presume it's another variation from the text of the article. If it's not, then please delete it altogether.

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/LibraryApplicationController.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        return bookRepository.get(title);
+    }
+    
+    @PostMapping("books/")
+    public void addBook(@RequestBody Book book) {
+        bookRepository.save(book);
+    }
+    
+    @Autowired
+    private LibrarianRepository librarianRepository;
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/LibraryApplicationController.java:69]: Fields should really be at the top of the class before the methods

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/LibraryApplicationController.java [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        return subscriberRepository.findById(id);
+    }
+    
+    @PostMapping("subscribers")
+    public Subscriber addSubscriber(@RequestBody Subscriber subscriber) {
+        return subscriberRepository.save(subscriber);
+    }
+    
+    @Autowired
+    private GatekeeperRepository gatekeeperRepository;
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/LibraryApplicationController.java:42]: Field split by methods

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/LibraryApplicationController.java [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+import com.baeldung.spring.data.redis_ttl.entity.Librarian;
+import com.baeldung.spring.data.redis_ttl.entity.Subscriber;
+import com.baeldung.spring.data.redis_ttl.repository.BookRepository;
+import com.baeldung.spring.data.redis_ttl.repository.GatekeeperRepository;
+import com.baeldung.spring.data.redis_ttl.repository.LibrarianRepository;
+import com.baeldung.spring.data.redis_ttl.repository.SubscriberRepository;
+
+@RestController
+@RequestMapping("/v1/")
+class LibraryApplicationController{
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/LibraryApplicationController.java:25]: Missing space before `{`

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/LibraryApplication.java [hunk 20]
```diff
@@ -0,0 +1,13 @@
+package com.baeldung.spring.data.redis_ttl;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.cache.annotation.EnableCaching;
+
+@EnableCaching
+@SpringBootApplication
+public class LibraryApplication {
+	public static void main(String[] args) {
```
Comment (ashleyfrieze) [inline persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/LibraryApplication.java:10]: This file needs to be indented with spaces, not tabs

**Review by ashleyfrieze — CHANGES_REQUESTED** [review]

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/repository/BookRepository.java (hunk 8 above)
Comment (sharanm) [reply persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/repository/BookRepository.java:31]: Java does allow unbraced if stmts, right ? So, is the comment more from a styling perspective?

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/LibraryApplicationController.java (hunk 18 above)
Comment (sharanm) [reply persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/LibraryApplicationController.java:42]: Sorry, didn't quite follow this.

File: persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/configuration/RedisConfig.java (hunk 16 above)
Comment (sharanm) [reply persistence-modules/spring-data-redis/src/main/java/com/baeldung/spring/data/redis_ttl/configuration/RedisConfig.java:47]: Correct, let me add comment.

File: persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java (hunk 6 above)
Comment (sharanm) [reply persistence-modules/spring-data-redis/src/test/java/com/baeldung/spring/data/redis_ttl/LibraryIntegrationTest.java:31]: Not really sure about this. I actually looked at other modules under spring-data-redis and used the same approach.
Any suggestion that you have to on how to test this ?

**Pull request closed without merging**